### PR TITLE
[SMTChecker] Change CHC encoding from explicit CFG to function forests

### PR DESCRIPTION
**Pull request by leonardoalt**
**tl;dr**
This PR changes the encoding in a way that is necessary for function calls, without implementing function calls just yet.
This PR is the longest in this chain, as implementing internal function calls after this one is easy.
I tried to keep at least some commits separated, but let me know if it's still too hard to review.

**Details:**

The current encoding is very similar to a contract's CFG, with the extra interface and error blocks.
As a small example, the following program would roughly have the presented encoding:
```
contract C {
    uint a;
    function f(uint x) public { assert(x > a); }
    function g(uint y) public { assert(y > a); }
}
```

```
true -> constructor
constructor -> interface(a)
interface(a) -> f_entry(a, x)
f_entry(a, x) -> f_assert(a, x)
f_assert(a, x) && !condition -> error_f
f_assert(a, x) && condition -> f_exit(a, x)
f_exit(a, x) -> interface
interface(a) -> g_entry(a, y)
... similar to f
```
We then query `error_f` and `error_g`.

However, this doesn't work for function calls, since we would either need to inline the whole CFG of the called function for every call, or we can't have more than one call to the same function, since we wouldn't know where to go back to.

The new encoding uses function summaries and nonlinear Horn clauses for that, where the function is encoded separately and "called" by the interface. Assertions do not create a new error block, they simply exit the current function with error. For that, a new error flag present in all predicates is added, where `error = 0` means no error, and `error > 0` means error where the value of `error` is the id of the failed assertion.

Besides the error flag, the main difference is the function summary predicate: It tells how a function behaves with respect to the contract state variables and function output variables in a summarized way:
`h_summary(error, s_0, s_1, in_0, in_1, out_0)`
This means that function `h` starts with state `s_0` and input `in_0`, and at the end of its execution state is `s_1` and it returns `out_1`.
`s_0` and `in_0` are simply passed from predicate to predicate unchanged so we know at any time how the execution started.

The program above would be roughly encoded as:

```
1. true -> genesis
2. genesis -> constructor_C(0)
3. constructor_C(error) && error == 0 -> interface(a)
4. genesis && a_0 == a_1 && x_0 == x_1 -> f_entry(0, a_0, a_1, x_0, x_1)
5. f_entry(error, a_0, a_1, x_0, x_1) && !(x_1 > a_1)  -> f_summary(1, a_0, a_1, x_0, x_1)
6. f_entry(error, a_0, a_1, x_0, x_1) -> f_exit(error, a_1, x_1)
7. f_exit(error, a_0, a_1, x_0, x_1) -> f_summary(error, a_0, a_1, x_0, x_1)
8. interface(a_0) && f_summary(error, a_0, a_1, x_0, x_1) && error == 0 -> interface(a_1)
9. interface(a_0) && f_summary(error, a_0, a_1, x_0, x_1) && error == 1 -> error_assertion_1
...similar for g...
```

We then query `error_assertion_1` and `error_assertion_2` (from `g`) separately.

Note that internal functions do not generate interface rules 8 and 9.
Also note that in rule 6, we do **not** add the assertion condition as a constraint in the function's path forward, so that we can still catch further assertion failures.

The changes made here should not change the results or support, but nonlinear Horn clauses make the system harder to solve, so we get some new false positives.

### CODE REVIEW

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -601,7 +601,12 @@ bool CHC::shouldVisit(ContractDefinition const& _contract) const
 
 bool CHC::shouldVisit(FunctionDefinition const& _function) const
 {
-	return _function.isImplemented();
+	if (
+		_function.isPublic() &&
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]: Not that it's too relevant, but: ``return _function.isPublic() && _function.isImplemented();`` :-)?

File: libsolidity/formal/CHC.h [hunk 2]
```diff
@@ -79,44 +79,76 @@ class CHC: public SMTEncoder
 	void unknownFunctionCall(FunctionCall const& _funCall);
 	//@}
 
+	struct IdCompare
```
Comment (ekpyron) [inline libsolidity/formal/CHC.h:82]: By the way I created https://github.com/ethereum/solidity/issues/8319 for tracking to collect all of these and maybe share the logic across the compiler.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `4be989acf346899b49b0070c9667a9cd09e1937f`:
```
Error: Trailing whitespace found:
 test/libsolidity/smtCheckerTests/functions/library_constant.sol:4: 

```
Please check that your changes are working as intended.

File: libsolidity/formal/CHC.cpp [hunk 3]
```diff
@@ -75,16 +77,42 @@ void CHC::analyze(SourceUnit const& _source)
...
 		vector<smt::SortPointer>(),
 		boolSort
 	);
 	m_genesisPredicate = createSymbolicBlock(genesisSort, "genesis");
-	auto genesis = (*m_genesisPredicate)({});
-	addRule(genesis, genesis.name);
+	addRule(genesis(), "genesis");
+
+	defineInterfacesAndSummaries(_source);
+	for (auto const& source: _source.referencedSourceUnits())
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]: Do you need to set the ``recursive`` flag of ``referencedSourceUnits`` to ``true``? Can this get circular in some weird cases? Also I think we should try to achieve a deterministic order here as well - some way down the call chain this will invoke ``uniquePrefix()``, which will end up non-deterministic, which might lead to weird hard-to-track non-determinism down the road, don't you think?

So maybe a create a ID-sorted set with ``_source`` in it and then add the recursive ``referencedSourceUnits(true)`` to it and only then walk that set calling ``defineInterfacesAndSummaries``?

File: libsolidity/formal/CHC.cpp [hunk 4]
```diff
@@ -75,16 +77,42 @@ void CHC::analyze(SourceUnit const& _source)
...
+		defineInterfacesAndSummaries(*source);
 
 	_source.accept(*this);
+	while (!m_toAnalyzeContracts.empty())
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]: Hm... where is this actually filled? Am I missing something or will this currently always be empty?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `d227aba4694551c9e0ac647181c28cfcb4a6282b`:
```
Error: Trailing whitespace found:
 test/libsolidity/smtCheckerTests/functions/library_constant.sol:4: 

```
Please check that your changes are working as intended.

File: libsolidity/formal/CHC.h (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.h:82]: Nice!

File: libsolidity/formal/CHC.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Why would it be nondet?

Sounds good anyway.

File: libsolidity/formal/CHC.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Currently it's empty. It's filled when there's a function call to a contract that is not `m_currentContract` or any of its bases.

File: libsolidity/formal/CHC.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Would you say that because if the referenced source units visit, the toAnalyzeContracts member is redundant?

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp]: The order of referenced source units is nondet, since it's a set of pointers. First I thought that might not be relevant, but ``defineInterfacesAndSummaries`` calls ``createSummaryBlock`` which calls ``uniquePrefix()`` to generate names, which is a simple counter. So you'll end up with non-deterministic names for the summaries and I'm pretty sure that way we loose all determinism-guarantees of the solvers.

File: libsolidity/formal/CHC.cpp (hunk 4 above)
Comment (ekpyron) [reply libsolidity/formal/CHC.cpp]: Ok - so it might make sense to remove it here and move it to the PR that introduces that... but I'd also be fine with keeping it here, if it'd be added by the next PR anyways.

File: libsolidity/formal/CHC.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Right. I applied your suggestion

**Comment by ekpyron** [comment]
Hm, this *looks* good now, but unfortunately we have to figure out where these memory corruptions are coming from...

File: libsolidity/formal/CHC.cpp [hunk 5]
```diff
@@ -75,16 +77,39 @@ void CHC::analyze(SourceUnit const& _source)
...
+	addRule(genesis(), "genesis");
+
+	set<SourceUnit const*, IdCompare> sources;
+	sources.insert(&_source);
+	for (auto const& source: _source.referencedSourceUnits(true))
+		sources.insert(source);
+	for (auto const* source: sources)
+		defineInterfacesAndSummaries(*source);
+	for (auto const* source: sources)
+		source->accept(*this);
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp:97]: Ah, yes, this should now really make the ``toAnalyze`` stuff you had earlier obsolete, now I agree.

Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:97]: exactly ;)

File: libsolidity/formal/CHC.cpp [hunk 6]
```diff
@@ -557,10 +604,50 @@ void CHC::setCurrentBlock(
 		m_currentBlock = predicate(_block);
 }
 
+set<Expression const*, CHC::IdCompare> CHC::transactionAssertions(ASTNode const* _txRoot)
+{
+	set<Expression const*, IdCompare> assertions;
+	set<ASTNode const*> visited;
+	queue<ASTNode const*> toVisit;
+	toVisit.push(_txRoot);
+	while (!toVisit.empty())
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]: You could use ``BreadthFirstSearch`` from ``libsolutil/Algorithms.h`` for this - the entire function would reduce to:
```
set<Expression const*, IdCompare> assertions;
BreadthFirstSearch<ASTNode const*>{{_txRoot}}.run([&](auto const* function, auto&& _addChild) {
  assertions.insert(m_functionAssertions[function].begin(), m_functionAssertions[function].end());
  for (auto const* called: m_callGraph[function])
    _addChild(called);
});
return assertions;
```

**Review by ekpyron — COMMENTED** [review]
Pending my last comment about reusing ``BreadthFirstSearch``, I think this is fine.

By the way: have you any plans to support function pointers, resp. function type variables? I think I have a working proposal for supporting them in mind... but we can discuss that next month.

File: libsolidity/formal/CHC.cpp (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Ah nice

**Comment by leonardoalt** [comment]
I don't have anything in mind re function pointers. I tend to give it very low priority since the easy way for us is not possible since Z3 doesn't allow functions as arguments (well, first order), and I don't think a lot of people use it. But if you have a proposal that works we should implement it.

**Comment by ekpyron** [comment]
Do you have an issue for it (re function pointers)? I didn't see one... I think it should work to just have an additional "internal dispatch" predicate for each function signature, which just takes an additional integer as argument which selects the particular internal function of that signature to be called (or the selector for external calls) - then you can just treat internal function pointers as integers (and external ones probably as pair of address and selector) and don't need actual functions as arguments and it should be fine even with Z3. If we have an issue for it, I can elaborate a bit more.

**Comment by leonardoalt** [comment]
@ekpyron https://github.com/ethereum/solidity/issues/8330

**Comment by leonardoalt** [comment]
Updated to use BFS

File: libsolidity/formal/CHC.cpp [hunk 7]
```diff
@@ -193,7 +211,17 @@ bool CHC::visit(FunctionDefinition const& _function)
...
+	if (_function.isConstructor())
+		connectBlocks(m_currentBlock, functionPred);
+	else
+		connectBlocks(genesis(), functionPred);
+
+	m_context.addAssertion(m_error.currentValue() == 0);
+	for (auto const* var: m_stateVariables)
+		m_context.addAssertion(m_context.variable(*var)->valueAtIndex(0) == currentValue(*var));
+	for (auto const& var: _function.parameters())
+		m_context.addAssertion(m_context.variable(*var)->valueAtIndex(0) == currentValue(*var));
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp:221]: Just so I get this right: you could also instead use ``connectBlock(genesis(), initialFunctionVariables())`` above instead and have ``initialFunctionVariables()`` be the same as ``currentFunctionVariables()`` just with init variables being the same as the "mutable" variables (i.e. the current values, not the zero-index values)?
And you instead add the assertions that they are the same here, so that you don't need to basically duplicate what ``currentFunctionVariables()`` does?

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:221]: And in general the only reason for the "function entry block" to even have two sets of arguments is so that every block of the function deals with the same set of variables?

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:221]: At least for the arguments... for the state variables the "initial state" will always be required for supporting function calls and reverting I guess. Seems like a fine concept to me, just making sure I get it correctly.

File: libsolidity/formal/CHC.cpp [hunk 8]
```diff
@@ -27,6 +27,10 @@
 
 #include <libsolidity/ast/TypeProvider.h>
 
+#include <libsolutil/Algorithms.h>
+
+#include <queue>
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]: This is unused in this latest version, isn't it?

**Review by ekpyron — COMMENTED** [review]
If I got the stuff I commented about before right and after removing the now superfluous header, I'm fine with merging.

File: libsolidity/formal/CHC.cpp (hunk 7 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:221]: Yes, that's the idea. Connecting `genesis` with the variables directly wouldn't work because then its sort would also need to contains the variables of the functions.

File: libsolidity/formal/CHC.cpp (hunk 8 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Removed

**Comment by leonardoalt** [comment]
Removed the useless header and rebased

**Review by ekpyron — APPROVED** [review]
